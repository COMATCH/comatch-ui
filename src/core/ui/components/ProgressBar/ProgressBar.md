## ProgressBar

prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**displayProgressLabel** | `Boolean` | `true` | :x: | 
**greyBackground** | `Boolean` | `false` | :x: | 
**label** | `String` | `''` | :x: | Label/title of the progress bar
**labelPosition** | `Enum('right', 'left')` | `'right'` | :x: | Position of the label/title of the progress bar
**progress** | `(custom validator)` | `0` | :x: | 
**progressLabel** | `String` | `''` | :x: | 
**progressiveColoring** | `Boolean` | `false` | :x: | 
**reverse** | `Boolean` | `false` | :x: | Progress bar from left to right
**variant** | `Enum('standard', 'light', 'dark')` | `'standard'` | :x: | Coloring variants of the progress bar and it's labels

