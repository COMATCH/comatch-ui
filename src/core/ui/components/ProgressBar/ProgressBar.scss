@import '../../../styles/variables.scss';

.ProgressBar {
    position: relative;
    height: 20px;
    border: solid 1px $co-color-mid-gray;
    border-radius: 3px;
    background-color: $co-color-white;

    &__progress {
        height: 100%;
        border-radius: 3px 0 0 3px;
        background-color: $co-color-primary;
        transform: scaleX(0);
        transform-origin: left;
        transition: transform 0.5s ease-out;

        &.full {
            border-radius: 3px;
        }
    }

    &__progress-label {
        font-size: $co-font-xs;
        position: absolute;
        top: 0;
        color: $co-color-primary;
        text-align: left;

        &.ProgressBar__progress-label-left {
            color: $co-color-white;
            text-align: right;
        }
    }

    // states
    &--secondaryColor {
        .ProgressBar__progress {
            background-color: $co-color-secondary;
        }
    }

    &--greyBackground {
        background-color: $co-color-mid-gray;
        border: 0;
    }

    &--danger {
        .ProgressBar__progress {
            background-color: $co-color-dark-red;
        }

        .ProgressBar__progress-label {
            color: $co-color-dark-red;
        }
    }

    &--warning {
        .ProgressBar__progress {
            background-color: $co-color-dark-orange;
        }

        .ProgressBar__progress-label {
            color: $co-color-dark-orange;
        }
    }
}
