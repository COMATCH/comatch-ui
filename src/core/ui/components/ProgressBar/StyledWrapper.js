import styled from 'styled-components';
import { palette, typography } from '../../../styles/variables';

export const StyledWrapper = styled.div`
    .ProgressBar {
        position: relative;
        height: 20px;
        border: solid 1px ${palette.midGray};
        border-radius: 3px;
        background-color: ${palette.white};

        &__label {
            color: ${palette.darkGray};
            text-align: right;

            &--left {
                text-align: left;
            }

            &--light {
                color: ${palette.midGray};
                font-weight: normal;
            }
        }

        &__progress {
            height: 100%;
            border-radius: 3px 0 0 3px;
            transform: scaleX(0);
            transition: transform 0.5s ease-out;
            transform-origin: left;
            background-color: ${palette.primary};

            &.reverse {
                transform-origin: right;
                border-radius: 0 3px 3px 0;
            }

            &.full {
                border-radius: 3px;
            }

            &.light {
                background-color: ${palette.lightSecondary};
            }

            &.dark {
                background-color: ${palette.secondary};
            }
        }

        &__progress-label {
            font-size: ${typography.fontXs};
            position: absolute;
            top: 0;
            color: ${palette.primary};
            text-align: left;

            &.ProgressBar__progress--label-light {
                color: ${palette.lightSecondary};
                font-weight: bold;
            }
            &.ProgressBar__progress--label-dark {
                color: ${palette.secondary};
                font-weight: bold;
            }

            &.ProgressBar__progress-label-left {
                color: ${palette.white};
                text-align: right;
                font-weight: bold;
            }
        }

        // states

        &--greyBackground {
            background-color: ${palette.midGray};
            border: 0;
        }

        &--danger {
            .ProgressBar__progress {
                background-color: ${palette.darkRed};
            }

            .ProgressBar__progress-label {
                color: ${palette.darkRed};
            }
        }

        &--warning {
            .ProgressBar__progress {
                background-color: ${palette.darkOrange};
            }

            .ProgressBar__progress-label {
                color: ${palette.white};
            }
        }
    }
`;
