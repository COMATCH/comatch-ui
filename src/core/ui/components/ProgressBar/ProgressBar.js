import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { InputLabel } from '../InputLabel';
import { StyledWrapper } from './StyledWrapper';

const propTypes = {
    displayProgressLabel: PropTypes.bool,
    greyBackground: PropTypes.bool,
    progress(props, propName) {
        const value = props[propName];
        if ((!value && value !== 0) || value < 0 || value > 100) {
            return new Error('StepProgressBar requires progress prop as number between 0 and 100');
        }
        return false;
    },
    /** Label/title of the progress bar */
    label: PropTypes.string,
    /** Position of the label/title of the progress bar */
    labelPosition: PropTypes.oneOf(['right', 'left']),
    progressiveColoring: PropTypes.bool,
    progressLabel: PropTypes.string,
    /** Progress bar from left to right */
    reverse: PropTypes.bool,
    secondaryColor: PropTypes.bool,
    /** Border radius 0px */
    squareEnds: PropTypes.bool,
    /** Colouring variants of the progress bar and it's labels */
    variant: PropTypes.oneOf(['primary', 'secondary']),
};

const defaultProps = {
    displayProgressLabel: true,
    greyBackground: false,
    label: '',
    labelPosition: 'right',
    progressiveColoring: false,
    progressLabel: '',
    reverse: false,
    secondaryColor: false,
    squareEnds: false,
    variant: 'primary',
};

function renderProgressLabel(displayProgressLabel, progressLabel, progress) {
    if (displayProgressLabel) {
        return progressLabel.length > 0 ? progressLabel : `${progress}%`;
    }

    return null;
}

export const ProgressBar = ({
    displayProgressLabel,
    greyBackground,
    label,
    labelPosition,
    variant,
    progress,
    progressiveColoring,
    progressLabel,
    reverse,
    secondaryColor,
    squareEnds,
}) => {
    const progressStyle = { transform: `scaleX(${progress / 100})` };

    // per design specs, label is displayed
    // - at the right of the progress limit for progress < 40,
    // - at the left of the progress limit for progress >= 40,
    const progressLabelOnLeft = progress >= 40;

    // styles for progress bar from right to left
    const progressLabelPrimaryStyle = progressLabelOnLeft
        ? { right: `calc(${100 - progress}% + 8px)` }
        : { left: `calc(${progress}% + 8px)` };

    // styles for progress bar from left to right
    const progressLabelReverseStyle = progressLabelOnLeft
        ? { right: `calc(${progress}% - 38px)` }
        : { left: `calc(${100 - progress}% - 38px)` };

    const progressLabelStyle = !reverse ? progressLabelPrimaryStyle : progressLabelReverseStyle;

    const classes = classNames('ProgressBar', {
        'ProgressBar--danger': progressiveColoring && progress <= 39,
        'ProgressBar--warning': progressiveColoring && progress >= 40 && progress <= 74,
        'ProgressBar--secondaryColor': secondaryColor,
        'ProgressBar--greyBackground': greyBackground,
        'ProgressBar__squareEnds--left': reverse && squareEnds,
        'ProgressBar__squareEnds--right': !reverse && squareEnds,
    });

    const progressClasses = classNames('ProgressBar__progress', {
        full: progress >= 100,
        reverse,
        squareEnds,
        secondary: variant === 'secondary',
    });

    const labelClasses = classNames('ProgressBar__label', {
        'ProgressBar__label--left': labelPosition === 'left',
        'ProgressBar__label--secondary': variant === 'secondary',
    });

    return (
        <StyledWrapper>
            <InputLabel text={label} className={labelClasses} />
            <div className={classes}>
                <div className={progressClasses} style={progressStyle} />
                <div
                    className={`ProgressBar__progress-label ${progressLabelOnLeft &&
                        'ProgressBar__progress-label-left'} ${variant === 'secondary' &&
                        'ProgressBar__progress--label-secondary'}`}
                    style={progressLabelStyle}
                >
                    {renderProgressLabel(displayProgressLabel, progressLabel, progress)}
                </div>
            </div>
        </StyledWrapper>
    );
};

ProgressBar.propTypes = propTypes;
ProgressBar.defaultProps = defaultProps;
