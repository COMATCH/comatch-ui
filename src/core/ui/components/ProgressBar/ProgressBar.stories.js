import React from 'react';
import Highlight from 'react-highlight';
import { storiesOf } from '@storybook/react';
import { ProgressBar } from './ProgressBar';

storiesOf('ProgressBar', module)
    .add('10% progress', () => (
        <>
            <ProgressBar progress={10} />
            <Highlight className="html">{`<ProgressBar progress={10} />`}</Highlight>
        </>
    ))
    .add('10% progress with label', () => (
        <>
            <ProgressBar progress={10} label="A label" />
            <Highlight className="html">{`<ProgressBar progress={10} label="A label" />`}</Highlight>
        </>
    ))
    .add('10% progress with label on the left', () => (
        <>
            <ProgressBar progress={10} label="A label" labelPosition="left" />
            <Highlight className="html">
                {`<ProgressBar progress={10} label="A label" labelPosition="left" />`}
            </Highlight>
        </>
    ))
    .add('10% progress and squareEnds', () => (
        <>
            <ProgressBar progress={10} label="A label" squareEnds />
            <Highlight className="html">{`<ProgressBar progress={10} label="A label" squareEnds />`}</Highlight>
        </>
    ))
    .add('10% progress reverse', () => (
        <>
            <ProgressBar progress={10} reverse />
            <Highlight className="html">{`<ProgressBar progress={10} reverse />`}</Highlight>
        </>
    ))
    .add('10% progress reverse and squareEnds', () => (
        <>
            <ProgressBar progress={10} reverse squareEnds />
            <Highlight className="html">{`<ProgressBar progress={10} reverse squareEnds />`}</Highlight>
        </>
    ))
    .add('10% progress secondary variant', () => (
        <>
            <ProgressBar progress={10} variant="secondary" />
            <Highlight className="html">{`<ProgressBar progress={10} variant='secondary' />`}</Highlight>
        </>
    ))
    .add('10% progress with label, secondary variant', () => (
        <>
            <ProgressBar progress={10} label="A label" variant="secondary" />
            <Highlight className="html">
                {`<ProgressBar progress={10} label="A label" variant="secondary" />`}
            </Highlight>
        </>
    ))
    .add('10% progress with label, secondary variant and reverse', () => (
        <>
            <ProgressBar progress={10} label="A label" variant="secondary" reverse />
            <Highlight className="html">
                {`<ProgressBar progress={10} abel="A label" variant="secondary" reverse />`}
            </Highlight>
        </>
    ))
    .add('10% progress with label, secondary variant and reverse, label on left', () => (
        <>
            <ProgressBar progress={10} label="A label" variant="secondary" reverse labelPosition="left" />
            <Highlight className="html">
                {`<ProgressBar progress={10} label="A label" variant="secondary" reverse labelPosition="left" />`}
            </Highlight>
        </>
    ))
    .add('50% progress', () => (
        <>
            <ProgressBar progress={50} />
            <Highlight className="html">{`<ProgressBar progress={50} />`}</Highlight>
        </>
    ))
    .add('50% progress squareEnds', () => (
        <>
            <ProgressBar progress={50} squareEnds />
            <Highlight className="html">{`<ProgressBar progress={50} squareEnds />`}</Highlight>
        </>
    ))
    .add('50% progress with label', () => (
        <>
            <ProgressBar progress={50} label="A label" />
            <Highlight className="html">{`<ProgressBar progress={50} label="A label" />`}</Highlight>
        </>
    ))
    .add('50% progress reverse', () => (
        <>
            <ProgressBar progress={50} reverse />
            <Highlight className="html">{`<ProgressBar progress={50} reverse />`}</Highlight>
        </>
    ))
    .add('50% progress reverse and squareEnds', () => (
        <>
            <ProgressBar progress={50} reverse squareEnds />
            <Highlight className="html">{`<ProgressBar progress={50} reverse squareEnds />`}</Highlight>
        </>
    ))
    .add('50% progress reverse, secondary variant', () => (
        <>
            <ProgressBar progress={50} reverse variant="secondary" />
            <Highlight className="html">{`<ProgressBar progress={50} reverse variant="secondary" />`}</Highlight>
        </>
    ))
    .add('50% progress reverse, secondary variant with label', () => (
        <>
            <ProgressBar progress={50} reverse variant="secondary" label="A label" />
            <Highlight className="html">
                {`<ProgressBar progress={50} reverse variant="secondary" label="A label" />`}
            </Highlight>
        </>
    ))
    .add('80% progress', () => (
        <>
            <ProgressBar progress={80} />
            <Highlight className="html">{`<ProgressBar progress={80} />`}</Highlight>
        </>
    ))
    .add('100% progress', () => (
        <>
            <ProgressBar progress={100} />
            <Highlight className="html">{`<ProgressBar progress={100} />`}</Highlight>
        </>
    ));
