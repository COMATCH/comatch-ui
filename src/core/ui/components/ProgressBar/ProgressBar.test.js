import React from 'react';
import { create } from 'react-test-renderer';
import { mount } from 'enzyme';

import { ProgressBar } from './ProgressBar';
import '../../../config/tests/setup';

const props = {
    displayProgressLabel: false,
    greyBackground: true,
    label: 'Label test',
    labelPosition: 'left',
    lightVersion: true,
    progress: 20,
    progressiveColoring: true,
    progressLabel: 'Test',
    reverse: true,
    secondaryColor: false,
    squareEnds: true,
};

describe('ProgressBar', () => {
    it('should render correctly', () => {
        const wrapper = create(<ProgressBar />).toJSON();
        expect(wrapper).toMatchSnapshot();
    });

    it('should render with default props', () => {
        const wrapper = mount(<ProgressBar />);

        expect(wrapper.prop('displayProgressLabel')).toEqual(true);
        expect(wrapper.prop('greyBackground')).toEqual(false);
        expect(wrapper.prop('label')).toEqual('');
        expect(wrapper.prop('labelPosition')).toEqual('right');
        expect(wrapper.prop('lightVersion')).toEqual(false);
        expect(wrapper.prop('progressiveColoring')).toEqual(false);
        expect(wrapper.prop('progressLabel')).toEqual('');
        expect(wrapper.prop('reverse')).toEqual(false);
        expect(wrapper.prop('secondaryColor')).toEqual(false);
        expect(wrapper.prop('squareEnds')).toEqual(false);
    });

    it('should render with custom props', () => {
        const wrapper = mount(<ProgressBar {...props} />);

        expect(wrapper.prop('displayProgressLabel')).toEqual(false);
        expect(wrapper.prop('greyBackground')).toEqual(true);
        expect(wrapper.prop('label')).toEqual('Label test');
        expect(wrapper.prop('labelPosition')).toEqual('left');
        expect(wrapper.prop('lightVersion')).toEqual(true);
        expect(wrapper.prop('progressiveColoring')).toEqual(true);
        expect(wrapper.prop('progressLabel')).toEqual('Test');
        expect(wrapper.prop('reverse')).toEqual(true);
        expect(wrapper.prop('secondaryColor')).toEqual(false);
        expect(wrapper.prop('progress')).toEqual(20);
        expect(wrapper.prop('squareEnds')).toEqual(true);
    });
});
