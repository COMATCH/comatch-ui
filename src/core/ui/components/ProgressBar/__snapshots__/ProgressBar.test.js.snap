// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar should render correctly 1`] = `
<div
  className="sc-bwzfXH bhjSjD"
>
  <label
    className="InputLabel ProgressBar__label sc-bdVaJa iThGRf"
    htmlFor={null}
    id={null}
  >
    
  </label>
  <div
    className="ProgressBar"
  >
    <div
      className="ProgressBar__progress"
      style={
        Object {
          "transform": "scaleX(NaN)",
        }
      }
    />
    <div
      className="ProgressBar__progress-label false false"
      style={
        Object {
          "left": "calc(undefined% + 8px)",
        }
      }
    >
      undefined%
    </div>
  </div>
</div>
`;
